### PR TITLE
feat(parachain/collator): handle Declare message received by a collator

### PR DESCRIPTION
**Pull request by kishansagathiya**
- Verify the collator by its signature and
- Save its collator id, peer id and Para id it collates for in our memory

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues
Fixes https://github.com/ChainSafe/gossamer/issues/3513

## Primary Reviewer

@timwu20

### CODE REVIEW

File: dot/parachain/collator-protocol/message.go [hunk 1]
```diff
@@ -175,32 +177,118 @@ const (
 	Seconded
 )
 
+// getDeclareSignaturePayload gives the payload that should be signed and included in a Declare message.
+// The payload is a the local peed id of the node, which serves to prove that it controls the
```
Comment (jimjbrettj) [inline dot/parachain/collator-protocol/message.go:181]:
```suggestion
// The payload is a the local peer id of the node, which serves to prove that it controls the
```

File: dot/parachain/collator-protocol/message.go [hunk 2]
```diff
@@ -175,32 +177,118 @@ const (
...
+	collatorProtocolMessageV, err := collatorProtocolMessage.Value()
+	if err != nil {
+		return propagate, fmt.Errorf("getting collator protocol message value: %w", err)
 	}
 
 	switch collatorProtocolMessage.Index() {
-	// TODO: Make sure that V1 and VStaging both types are covered
-	// All the types covered currently are V1.
+	// TODO: Make sure that V1 types are covered.
+	// TODO: Create an issue to cover v2 types.
```
Comment (jimjbrettj) [inline dot/parachain/collator-protocol/message.go:222]: is there an issue for this?

File: dot/parachain/collator-protocol/message_test.go [hunk 3]
```diff
@@ -172,3 +178,222 @@ func TestDecodeCollationHandshake(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, testHandshake, msg)
 }
+
+func TestHandleCollationMessageCommon(t *testing.T) {
```
Comment (jimjbrettj) [inline dot/parachain/collator-protocol/message_test.go:182]: is there a happy path to test here?

**Review by jimjbrettj — COMMENTED** [review]
Looks good, just a few comments!

File: dot/parachain/collator-protocol/message_test.go (hunk 3 above)
Comment (kishansagathiya) [reply dot/parachain/collator-protocol/message_test.go:182]: these are common errors that could happen. Happy path would change as per the message. There is a success case in TestHandleCollationMessageDeclare which is the happy path for that message.

File: dot/parachain/collator-protocol/message.go (hunk 2 above)
Comment (kishansagathiya) [reply dot/parachain/collator-protocol/message.go:222]: Just created one https://github.com/ChainSafe/gossamer/issues/3534
Will add more details into it

File: dot/parachain/collator-protocol/message.go [hunk 4]
```diff
@@ -175,32 +177,118 @@ const (
...
```
Comment (kishansagathiya) [inline dot/parachain/collator-protocol/message.go:222]:
```suggestion
	// TODO: Create an issue to cover v2 types. #3534
```

File: dot/parachain/collator-protocol/message_test.go [hunk 5]
```diff
@@ -172,3 +178,226 @@ func TestDecodeCollationHandshake(t *testing.T) {
...
+		c := c
+		t.Run(c.description, func(t *testing.T) {
+			t.Parallel()
+			cpvs := CollatorProtocolValidatorSide{
+				net:                c.net,
+				peerData:           c.peerData,
+				currentAssignments: c.currentAssignments,
+			}
+			msg := NewCollationProtocol()
+			vdt_child := NewCollatorProtocolMessage()
```
Comment (axaysagathiya) [inline dot/parachain/collator-protocol/message_test.go:379]: Use camel case instead
```suggestion
			vdtChild := NewCollatorProtocolMessage()
```

**Review by jimjbrettj — APPROVED** [review]
lgtm!

**Review by axaysagathiya — APPROVED** [review]

**Pull request merged**